storage/src/main/java/tech/pegasys/teku/storage/store/Store.java

```diff
@@ -461,10 +471,64 @@ public SafeFuture<Optional<SignedBeaconBlock>> retrieveSignedBlock(final Bytes32
   }
 
   @Override
-  @SuppressWarnings("unused")
   public SafeFuture<Optional<SignedBeaconBlockAndBlobsSidecar>> retrieveSignedBlockAndBlobsSidecar(
       final Bytes32 blockRoot) {
-    return SafeFuture.failedFuture(new UnsupportedOperationException("Not yet implemented"));
+
+    return retrieveSignedBlock(blockRoot)
+        .thenCompose(
+            signedBeaconBlock -> {
+              if (signedBeaconBlock.isEmpty()) {
+                return SafeFuture.completedFuture(Optional.empty());
+              }
+              return retrieveSignedBlockAndBlobsSidecar(signedBeaconBlock.get())
+                  .thenApply(Optional::of);
+            });
+  }
+
+  private SafeFuture<SignedBeaconBlockAndBlobsSidecar> retrieveSignedBlockAndBlobsSidecar(
+      final SignedBeaconBlock signedBeaconBlock) {
+    return blobsSidecarProvider
+        .getBlobsSidecar(signedBeaconBlock)
+        .thenApply(
+            blobsSidecar ->
+                createSignedBeaconBlockAndBlobsSidecar(signedBeaconBlock, blobsSidecar));
+  }
+
+  private SignedBeaconBlockAndBlobsSidecar createSignedBeaconBlockAndBlobsSidecar(
+      final SignedBeaconBlock signedBeaconBlock, final Optional<BlobsSidecar> maybeBlobsSidecar) {
+    checkState(
+        maybeBlobsSidecar
+            .map(
+                blobsSidecar ->
+                    blobsSidecar.getBeaconBlockRoot().equals(signedBeaconBlock.getRoot()))
+            .orElseGet(
+                () ->
+                    signedBeaconBlock
+                        .getBeaconBlock()
+                        .orElseThrow()
+                        .getBody()
+                        .toVersionEip4844()
+                        .orElseThrow()
+                        .getBlobKzgCommitments()
+                        .isEmpty()),
+        "BlobsSidecar is inconsistent with the BeaconBlock");
```

Comment:
I would suggest to add the reason of the inconsistency (either the `BlobsSidecar` couldn't be found at all or the block root doesn't match)

Reply:
point is that I mixed two different conditions here. Btw it shouldn't happen because the data comes from our own db so it is more our own consistency check instead of a validation. Btw I can split it in two, so it can be more readable.